### PR TITLE
Add support for ClusterDocsTopics in helm-broker

### PR DESCRIPTION
**Pull request by polskikiel**
**Description**

Changes proposed in this pull request:

- Add support for ClusterDocsTopics in helm-broker

**testing**

1. Use bundles from this release: 
```
https://github.com/polskikiel/bundles/releases/download/latest/index.yaml
```

2. Use the following version of helm broker:
```
eu.gcr.io/kyma-project/pr/helm-broker:PR-3769
```
3. Add the following RBAC roles to Helm broker:
```
- apiGroups:
  - cms.kyma-project.io
  resources:
  - clusterdocstopics
  verbs:
  - get
  - update
  - delete
  - create
```

**Related issue(s)**

### CODE REVIEW

**Comment by PK85** [comment]
Please adjust two things I found in the https://github.com/kyma-project/kyma/pull/3751/files:
1) remote Remote from RemoteRepositoryURL (only go code refactoring)
2) Helm Values key, from :"Values.bundleRepositoryURL" to "Values.addonRepositoryURL"

**Comment by polskikiel** [comment]
@PK85 
Ok, adjusted.

File: components/helm-broker/internal/bundle/form.go [hunk 1]
```diff
@@ -36,6 +37,11 @@ type FormMeta struct {
 	PlanUpdatable       *bool             `yaml:"planUpdatable"`
 }
 
+// DocsMeta contains data about bundle's docs fetched from docs/meta.yaml file
+type DocsMeta struct {
+	Docs []internal.BundleDocs `yaml:"docs"`
```
Comment (mszostok) [inline components/helm-broker/internal/bundle/form.go]: you should have a `Validate` method which will ensure that there is no more that one entry

File: components/helm-broker/internal/bundle/syncer.go [hunk 2]
```diff
@@ -59,17 +71,32 @@ func (s *Syncer) Execute() error {
...
 	// remove bundles before upsert
 	if err := s.bundleStorage.RemoveAll(); err != nil {
 		return errors.Wrap(err, "while removing all bundles")
 	}
 
-	// upsert bundles
 	for _, bundleWithCharts := range uniqueBundles {
-		if err := s.upsert(bundleWithCharts); err != nil {
+		// already we support only single spec entry
+		if len(bundleWithCharts.Bundle.Docs) > 0 {
```
Comment (mszostok) [inline components/helm-broker/internal/bundle/syncer.go]:
```suggestion
		if len(bundleWithCharts.Bundle.Docs) == 0 {
```

File: components/helm-broker/internal/bundle/syncer.go [hunk 3]
```diff
@@ -121,15 +153,80 @@ func (s *Syncer) removeAndNotifyAboutDuplication(allBundles []CompleteBundle) []
...
+		if err := s.ensureDocsTopicsRemoved(string(id)); err != nil {
+			return nil, errors.Wrapf(err, "while ensuring ClusterDocsTopic %s is removed", id)
+		}
+	}
+
+	return uniqueBundles, nil
+}
+
+func (s *Syncer) ensureDocsTopics(bundle CompleteBundle) error {
+	if err := s.ensureDocsTopicsRemoved(string(bundle.ID())); err != nil {
```
Comment (mszostok) [inline components/helm-broker/internal/bundle/syncer.go]: you should not remove the CDT. You should just create a CDT if `IsAlreadyExist` error is returned then execute Update.  Deleting and Creating is a different flow that Create or Updating. The CMS controller will work in totally different manner

File: components/helm-broker/internal/bundle/docs_provider.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package bundle
+
+import (
+	"context"
+
+	"github.com/kyma-project/kyma/components/cms-controller-manager/pkg/apis/cms/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/bundle/docs_provider.go]: done

File: components/helm-broker/internal/bundle/docs_provider.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// DocsProvider allows to maintain the bundles documentation
+type DocsProvider struct {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/bundle/docs_provider.go]: discussed, I left this structure and moved the ensuring logic into it.

File: components/helm-broker/internal/bundle/form.go (hunk 1 above)
Comment (polskikiel) [reply components/helm-broker/internal/bundle/form.go]: added

File: components/helm-broker/internal/bundle/syncer.go [hunk 6]
```diff
@@ -59,17 +71,33 @@ func (s *Syncer) Execute() error {
...
 	}
 
-	// upsert bundles
+	// upsertBundle bundles
 	for _, bundleWithCharts := range uniqueBundles {
-		if err := s.upsert(bundleWithCharts); err != nil {
+		// already we support only single spec entry
+		if len(bundleWithCharts.Bundle.Docs) > 0 {
+			if err := s.ensureDocsTopics(bundleWithCharts); err != nil {
+				s.log.Errorf("While ensuring docs topic for bundle %s: %v", bundleWithCharts.ID(), err)
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/bundle/syncer.go]: fixed

File: components/helm-broker/internal/bundle/syncer.go [hunk 7]
```diff
@@ -59,17 +69,29 @@ func (s *Syncer) Execute() error {
...
 
-	fetchedBundles := s.fetchBundlesFromRepositories()
-	uniqueBundles := s.removeAndNotifyAboutDuplication(fetchedBundles)
+	previousBundles, err := s.bundleStorage.FindAll()
+	if err != nil {
+		s.log.Errorf("Could not load existing bundles, error: %s", err)
+	}
+	newBundles := s.fetchBundlesFromRepositories()
+	if err := s.deleteUnusedDocsTopics(previousBundles, newBundles); err != nil {
+		s.log.Errorf("while deleting unused ClusterDocsTopics: %v", err)
```
Comment (mszostok) [inline components/helm-broker/internal/bundle/syncer.go]: why we do not want to return the error?

File: components/helm-broker/internal/bundle/syncer.go [hunk 8]
```diff
@@ -59,17 +69,29 @@ func (s *Syncer) Execute() error {
 	s.log.Info("Loading bundles from repositories")
 	defer s.log.Info("Loading bundles finished")
 
-	fetchedBundles := s.fetchBundlesFromRepositories()
-	uniqueBundles := s.removeAndNotifyAboutDuplication(fetchedBundles)
+	previousBundles, err := s.bundleStorage.FindAll()
+	if err != nil {
+		s.log.Errorf("Could not load existing bundles, error: %s", err)
```
Comment (mszostok) [inline components/helm-broker/internal/bundle/syncer.go]: why we do not want to return the error?

File: components/helm-broker/internal/bundle/syncer.go [hunk 9]
```diff
@@ -121,15 +147,47 @@ func (s *Syncer) removeAndNotifyAboutDuplication(allBundles []CompleteBundle) []
...
 
-	return uniqueBundles
+	return uniqueBundles, nil
+}
+
+func (s *Syncer) deleteUnusedDocsTopics(previousBundles []*internal.Bundle, newBundles []CompleteBundle) error {
+	// calculate removed
+	var idsToRemove []internal.BundleID
+
+	for _, v := range previousBundles {
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/bundle/syncer.go]: Ok, so I changed the whole Syncer to make it `O(n)`, I moved some logic from `removeAndNotifyAboutDuplication` to `fetchBundlesFromRepositories`. Now it looks like this: 
```
// collect results
	fetchedBundles := make(map[internal.BundleID][]CompleteBundle)
	for range s.repositories {
		change := <-resultChan
		if !change.containsChange() {
			continue
		}

		for _, item := range change.items {
			if _, exists := fetchedBundles[item.ID()]; !exists {
				fetchedBundles[item.ID()] = []CompleteBundle{}
			}
			fetchedBundles[item.ID()] = append(fetchedBundles[item.ID()], item)
		}
	}

	return fetchedBundles
```
And then I have a map to decrease the complexity level in this function
```
func (s *Syncer) deleteUnusedDocsTopics(previousBundles []*internal.Bundle, newBundles map[internal.BundleID][]CompleteBundle) error {
	for _, v := range previousBundles {
		if _, exists := newBundles[v.ID]; !exists {
			if err := s.docsTopicsService.EnsureClusterDocsTopicRemoved(string(v.ID)); err != nil {
				return errors.Wrapf(err, "while ensuring ClusterDocsTopic %s is removed", v.ID)
			}
		}
	}
	return nil
```

File: components/helm-broker/internal/model.go [hunk 10]
```diff
@@ -127,6 +128,11 @@ func (b BundlePlanMetadata) ToMap() map[string]interface{} {
 // BundleTag is a Tag attached to Bundle.
 type BundleTag string
 
+// BundleDocs contains data to create ClusterDocsTopic for every ClusterServiceClass.
+type BundleDocs struct {
+	Template cms.ClusterDocsTopicSpec `yaml:"template"`
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/model.go]: deleted, yes i know but it was leftover from my fight with json parsing :)

File: components/helm-broker/cmd/broker/main.go [hunk 11]
```diff
@@ -12,21 +12,24 @@ import (
 	"time"
 
 	scCs "github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset"
-	"github.com/sirupsen/logrus"
-	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
-
+	"github.com/kyma-project/kyma/components/cms-controller-manager/pkg/apis/cms/v1alpha1"
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/cmd/broker/main.go]: they are already sorted by goimports, I tried to do it better but the goimport check failed. I will leave the goimport version

File: components/helm-broker/internal/bundle/docs_provider.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+		dynamicClient: dynamicClient,
+	}
+}
+
+// EnsureClusterDocsTopic creates ClusterDocsTopic for a given bundle or updates it in case it already exists
+func (d *DocsProvider) EnsureClusterDocsTopic(bundle *internal.Bundle) error {
+	bundle.Docs[0].Template.Sources = d.defaultBundleDocsSourcesURLs(bundle)
+
+	cdt := &v1alpha1.ClusterDocsTopic{
+		TypeMeta: v1.TypeMeta{
```
[the remark this answers is not in the document]
Comment (polskikiel) [reply components/helm-broker/internal/bundle/docs_provider.go]: I tested if that works without TypeMeta specified and it does. I did need to specify that when working with Applications etc. maybe the `k8s.io/code-generator` version differs. Deleted.

File: components/helm-broker/cmd/broker/main.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mszostok) [reply components/helm-broker/cmd/broker/main.go]: by saying please sort I mean to change the groups. as you can see we have 3 groups "core", "internal", "3rd party", but let's leave it and we can discuss this next time :)

**Pull request merged**